### PR TITLE
Add LoRA support for Gemma

### LINKED ISSUE #3044: Support LoRA for Gemma
I am getting the following error when loading a lora adaptor for Gemma model:

```bash
  python -u -m vllm.entrypoints.openai.api_server \
    --host 0.0.0.0 \
    --model google/gemma-7b \
    --enable-lora \
    --lora-modules quote=/artifacts \
    --max-model-len 1024 | tee ~/openai_api_server.log
```

```
ValueError: Model GemmaForCausalLM does not support LoRA, but LoRA is enabled. Support for this model may be added in the future. If this is important to you, please open an issue on github.
```

### PR DESCRIPTION
**Pull request by WoosukKwon**
Closes #3044

### CODE REVIEW

**Comment by Yard1** [comment]
Can we extend the test?

**Comment by WoosukKwon** [comment]
@Yard1 Added. PTAL!

**Review by Yard1 — APPROVED** [review]

File: csrc/punica/bgmv/bgmv_config.h [hunk 1]
```diff
@@ -28,6 +28,7 @@ void bgmv_kernel(out_T *__restrict__ Y, const in_T *__restrict__ X,
     f(in_T, out_T, W_T, narrow, 5120) \
     f(in_T, out_T, W_T, narrow, 5504) \
     f(in_T, out_T, W_T, narrow, 5632) \
+    f(in_T, out_T, W_T, narrow, 6144) \
```
Comment (Yard1) [inline csrc/punica/bgmv/bgmv_config.h:31]: let's also add those dimensions to the punica test?

Comment (WoosukKwon) [reply csrc/punica/bgmv/bgmv_config.h:31]: Sounds good! Added.

**Comment by WoosukKwon** [comment]
@Yard1 The PR continues to fail in `test_llama_lora_warmup` while this PR seems to be irrelevant to it. Do you have any idea about this? I tried to debug but didn't succeed.

**Comment by Yard1** [comment]
Hmm running the entire thing as a new process should help with cleanup

**Comment by WoosukKwon** [comment]
@Yard1 Adding `--forked` indeed resolved the issue but increased the lora test time from 13 mins to 37 mins.

**Comment by WoosukKwon** [comment]
Let's merge the PR and optimize the testing time later. I thin this is ok since the lora test is not required for most PRs and the kernel test (which also uses `--forked`) takes a similar amount of time.

**Comment by zhaotyer** [comment]
Gemma "vocab_size" is 256000, Will the following restrictions have any impact? Should they be removed?
vllm/lora/layers.py
```
# Keep this in sync with csrc/punica/bgmv/bgmv_config.h
if 32000 < self.base_layer.vocab_size > 33024:
       raise ValueError(
                "When using LoRA, vocab size must be 32000 >= vocab_size <= 33024"
            )
```

**Comment by Yard1** [comment]
We are not supporting lm_head deltas for gemma

**Comment by zhaotyer** [comment]
> We are not supporting lm_head deltas for gemma

Thank you for your reply @Yard1 
```
            new_module = replace_submodule(
                            self.model, module_name,
                            from_layer(module, self.lora_slots, self.lora_config,
                                       self.model.config))
            # (yard1): TODO make this more robust
            if "lm_head" in module_name:
                sampler_module = self.model.get_submodule("sampler")
                new_module = replace_submodule(
                    self.model, "sampler",
                    from_layer_sampler(sampler_module, module, self.lora_slots,
                                       self.lora_config, self.model.config))
```
Vocal_size verification will only be performed when lm_head is also processed by lora, but please take a look at this question. #3000

**Pull request merged**